include/common/arch/mips/asm/utils.h

```diff
@@ -0,0 +1,54 @@
+#ifndef __UTILS_H__
+#define __UTILS_H__
+
+
+/*new file :from kernel/mips
+* fixme: gysun
+*
+*/
+
+#ifdef CONFIG_SMP
+# define kernel_uses_llsc	1
+#else
+# define kernel_uses_llsc	0
+#endif
+
+typedef struct {
+	int counter;
+}atomic_t;
+
+/*from kernel/linux-3.10.84/arch/mips/include/barrier.h*/
+#if defined(CONFIG_WEAK_ORDERING) && defined(CONFIG_SMP)
+# ifdef CONFIG_CPU_CAVIUM_OCTEON
+#  define smp_mb()	__sync()
```

Comment:
I think all the ifdefery with kernel config options can be removed as we cannot rely on them in userspace.
I'd choose the most restrictive implementation of the barrier primitives so that they would work correctly on any MIPS variant even at expense of performance